Billing models

dylanlott opened the pull request:
Initial pull request for billing models. 

I removed testing references, and included only the model changes and new GraphQL updates. 

I can add testing changes back in if wanted, but I think those would be fitting on a separate branch.

dylanlott committed:
heroku integration changes

dylanlott committed:
Some updates

dylanlott committed:
merge in storj master branch

bryanchriswhite committed:
Merge branch 'master' into billing-models

* master:
  fix npm link dep in package.json
  add npm link deps
  add new mirror model
  bump

Conflicts:
	package.json

bryanchriswhite committed:
Merge branch 'master' of github.com:Storj/service-storage-models into billing-models

* 'master' of github.com:Storj/service-storage-models:
  Updating readme to reflect correct way to instantiate connection
  major version bump just in case...
  updating merge
  bumping version
  Updates for correct config parsing
  updates for parsing configs and passing to mongoose allowing for configurable options

Conflicts:
	index.js

bryanchriswhite committed:
Merge branch 'master' of github.com:Storj/service-storage-models into billing-models

* 'master' of github.com:Storj/service-storage-models:
  bump to 4
  upgrade to storjlib5
  include missing a npm link dep

Conflicts:
	package.json

bryanchriswhite committed:
ლ(ಠ益ಠ)ლ

phutchins committed:
merging awesome changes

phutchins committed:
bleh

bryanchriswhite committed:
(>ლ)

bryanchriswhite committed:
Merge branch 'billing-models' of github.com:Storj/service-storage-models into billing-models

* 'billing-models' of github.com:Storj/service-storage-models:
  bleh

Commit:
WIP cleaned up bandwidth and storage model entries

Commit:
Merge branch 'billing-models' of github.com:Storj/service-storage-models into billing-models
WIP getting shit working

Commit:
Updated debit model, need to add back in usage metrics

barbaraliau committed:
Upgrade dependencies

Commit:
Edited bandwidth models in Debit

barbaraliau committed:
Initial

barbaraliau committed:
Merge branch 'billing-models' of github.com:Storj/service-storage-models into billing-models

Commit:
Merge branch 'master' of github.com:Storj/service-storage-models into billing-models-testing

barbaraliau committed:
Adding tests

barbaraliau committed:
Merge from upstream

barbaraliau committed:
Pass test for creating credit with the default props

barbaraliau committed:
Add string defaults for PROMO_CODES and PAYMENT_PROCESSORS

barbaraliau committed:
Add .pre('validate') funciton and change from arrow fns to regular fns

barbaraliau committed:
Remove merge refs

barbaraliau committed:
Passing Storage/models/Credit 'paid_amount' tests

barbaraliau committed:
Change DEFAULT to NONE for PROMO_CODES

barbaraliau committed:
Add more validation

barbaraliau committed:
Planned tests completed

barbaraliau committed:
Add additional validation

barbaraliau committed:
Make test name more specific

barbaraliau committed:
Passing tests. Need to review other tests to see if they are necessary

barbaraliau committed:
Made default test...not sure what else to test

barbaraliau committed:
Remove comments

Commit:
adding yarn.lock

Commit:
Merge branch 'billing-5' into billing-models

Commit:
adding yarn.lock

barbaraliau committed:
Passing tests checking create, lock, and unlock methods

barbaraliau committed:
Remove stray console.log

barbaraliau committed:
Organize into own describe('#method') blocks

barbaraliau committed:
Add additional validation

barbaraliau committed:
Formatting

barbaraliau committed:
Tests for create method

bryanchriswhite committed:
Merge remote-tracking branch 'storj/master' into billing-models

* storj/master: (45 commits)
  point to published lib6
  point to bookchins fork
  bump storj-lib6
  Bump to v5.1.5
  Use only user for index
  Add additional indexes
  Bump package.json to 5.1.3
  Add index
  bump
  Use exchange start and end timestamps instead of elapsed time
  bump travis version
  another fix
  fix booboo - bump
  bump to v5.1.0
  Use hyphen in file name
  Add ExchangeReport module to index file
  Update field names for consistency
  Rename file to match new collection name
  Change collection name to ExchangeReport
  Fix spacing
  ...

Conflicts:
	index.js
	package.json

bryanchriswhite committed:
wip systemd timer/unit for debit generation

bryanchriswhite committed:
Merge branch 'billing-models' of github.com:bryanchriswhite/service-storage-models into billing-models

* 'billing-models' of github.com:bryanchriswhite/service-storage-models:

Conflicts:
	lib/models/user.js
	package.json

barbaraliau committed:
Add _id:false to models and remove unnecessary code as a result of change. Refactor.

barbaraliau committed:
Shard.create tests

barbaraliau committed:
Add 'toObject' transform function to delete unnecessary fields

barbaraliau committed:
Add test for create method

barbaraliau committed:
Test for usernonce create method

barbaraliau committed:
Merge branch 'billing-models' of github.com:Storj/service-storage-models into billing-models

Commit:
WIP

Commit:
Merge branch 'billing-models' of github.com:bryanchriswhite/service-storage-models into billing-models

barbaraliau committed:
Delete yarn.lock

barbaraliau committed:
Uncommented to prevent payment info leak

barbaraliau committed:
Remove unnecessary method

barbaraliau committed:
Revert error codes

barbaraliau committed:
Fix typo

barbaraliau committed:
Rescue lost semi-colon

barbaraliau committed:
Fix whitespace

Commit:
Added graphQL to package.json

Commit:
Merge branch 'master' of github.com:Storj/service-storage-models into billing-models

barbaraliau committed:
Add graphql to dependencies

barbaraliau committed:
Revert back. Call .toObject() on shard otherwise transform function doesn't get called

barbaraliau committed:
Update shard tests to reflect changes in shard schema

barbaraliau committed:
Change variable to env var

barbaraliau committed:
Clean up code and TODOs

barbaraliau committed:
Merge with changes from upstream

barbaraliau committed:
Fix duplicate require

Commit:
Updates to billing models for remove card functionality

Commit:
Adds additional handling for card removal and re-adding

Commit:
WIP updates for remove card and multiple card handling

Commit:
WIP need to remove graphQL types from here

Commit:
Cleaning up for pull request

Commit:
Cleaning up for pull request

Commit:
Merge branch 'billing-models' of github.com:dylanlott/service-storage-models into billing-models

Commit:
Merge branch 'billing-models' of github.com:dylanlott/service-storage-models into billing-models

Commit:
Merge branch 'billing-models' of github.com:dylanlott/service-storage-models into billing-models

braydonf commented:
Notes from what were discussed at noon:

1) Seperate PaymentProcess schema out of User and to have a reference to user.
2) GraphQL dependency and code is not needed here, as it's duplicated in https://github.com/Storj/billing

Also, in reference to the testing. I didn't mean to remove ALL the tests from this code, but that there were tests included that was not related to these changes, and we could merge those independently.

bryanchriswhite committed:
upstream merge, cleanup, code-style:

+ move updated payment-processor-adapters from billing
+ remove unnecessary graphql stuff
+ cleanup / comment / code-style

bryanchriswhite committed:
questionable merge...

dylanlott committed:
Merge pull request #1 from bryanchriswhite/billing-models

Billing models

braydonf commented:
There are some merge conflicts

braydonf commented on lib/models/mirror.js:
```diff
@@ -3,6 +3,7 @@
 const errors = require('storj-service-error-types');
 const mongoose = require('mongoose');
 const SchemaOptions = require('../options');
+const _ = require('lodash');
```
We probably do not need to bring in lodash specifically for `_.isEmpty`, and lodash is not included in package.json, so this will likely fail.

barbaraliau replied:
Hmm..lodash was there when I wrote the test which is why I used `_.isEmpty`...guess it must have been put it from another merge somewhere. Test can be rewritten to not use lodash.

braydonf replied:
Part of the reason I had thought to have this be a separate PR,  since I don't think this is related to billing, but if you want to include with this one please just coordinate w/ dylan.

braydonf commented on lib/models/mirror.js:
```diff
@@ -38,13 +43,27 @@ Mirror.set('toObject', {
-  mirror.save(callback);
+  mirror.save(function(err, mirror) {
+    if (err) {
+      return callback(new Error('Cannot save mirror', err));
```
Why was this added?

barbaraliau replied:
Every other model that does a `.save()` follows this pattern. Is there a reason why `mirror.save` wouldn't?

braydonf replied:
Hmm.... I didn't realize that.

braydonf replied:
Actually it looks like this is common:
```js
shard.save(function(err) {
  if (err) {
    return callback(err);
  }
  callback(null, shard);
});
```

Is shard not returned back as the second argument of save? And likewise in this case mirror? If it is, then it can just be, as it was before:
```js
mirror.save(callback);
```

barbaraliau replied:
Mongoose's .save method returns the err and then the newly saved doc as the second argument. I'm not really sure why it's not returned in the callback.

braydonf replied:
I see, we should be fine to use `mirror.save(callback)` then.

braydonf commented on package.json:
```diff
@@ -28,16 +28,17 @@
   "homepage": "https://github.com/Storj/service-storage-models#readme",
   "dependencies": {
     "elliptic": "^6.3.2",
+    "graphql": "^0.8.2",
```
Not needed here.

barbaraliau commented on lib/models/credit.js:
```diff
@@ -5,18 +5,24 @@ const SchemaOptions = require('../options');
+const PROMO_CODES = constants.PROMO_CODES;
 
-// TODO: standardize property names to lowerCamelCase
+// NB: Check this shit out. Y no pass in earlier?
```
Delete this comment

braydonf commented on lib/models/user.js:
```diff
@@ -89,14 +186,29 @@ User.set('toObject', {
     delete ret.resetter;
     delete ret.pendingHashPass;
-    delete ret.bytesUploaded;
+    // delete ret.paymentProcessors;
```
I think these may have been already uncommented before, these should probably not be on the toObject for user?

barbaraliau replied:
Yes, these were uncommented

braydonf commented:
In regards to `calculateUserAmountOwedForBandwidth`, the reporterId can be both email and public key. However, I wouldn't worry about it now, as we're going to need a better way to calculate these values, since it will be quite easy to cheat the calculations. I have some ideas here, as we're currently tracking bandwidth for rate limiting purposes.

Commit:
Add dotenv environment variable file handling

Commit:
Merge branch 'master' of github.com:Storj/service-storage-models into billing-models

Commit:
Merge branch 'billing-models' of github.com:dylanlott/service-storage-models into billing-models

braydonf commented:
Looks like there is a missing dependency for the tests, and why it's failing.

braydonf commented on lib/models/user.js:
```diff
@@ -6,6 +6,129 @@ const activator = require('hat').rack(256);
+});
+
+PaymentProcessor.methods.refresh = function() {
+  console.log('USER REFRESHED.');
```
Stray console log?

braydonf commented on lib/models/user.js:
```diff
@@ -6,6 +6,129 @@ const activator = require('hat').rack(256);
+  return UserModel
+    .findOne({ _id: this.__parent._id })
+    .then((user) => {
+      console.log('USER MODEL REFRESH: ', user.defaultPaymentProcessor.data.customer.sources.data);
```
Another

braydonf commented on lib/models/user.js:
```diff
@@ -6,6 +6,129 @@ const activator = require('hat').rack(256);
+        this.__parent.paymentProcessors.length,
+        user.paymentProcessors
+      )
+      console.log('__PARENT', this.__parent.paymentProcessors);
```
Another

braydonf commented on lib/models/user.js:
```diff
@@ -294,5 +461,5 @@ User.statics.lookup = function(email, passwd, callback) {
 
 module.exports = function(connection) {
-  return connection.model('User', User);
+  return (UserModel = connection.model('User', User));
```
What's up here?

braydonf commented on package.json:
```diff
@@ -27,17 +27,19 @@
   "dependencies": {
+    "dotenv": "^2.0.0",
     "elliptic": "^6.3.2",
+    "graphql": "^0.8.2",
```
Not needed here.

braydonf commented on lib/models/payment-processor-adapters/stripe.js:
```diff
@@ -0,0 +1,201 @@
+   * @return {Boolean}
+   */
+  validate: function() {
+    console.log('stripeProcessor validate: ', stripeProcessor);
```
Another stray console log

braydonf commented on lib/models/payment-processor-adapters/stripe.js:
```diff
@@ -0,0 +1,201 @@
+    // TODO: verify with stripe that subscription/customer are still valid, etc.
+    const subscriptions = stripeProcessor.data.customer.subscriptions;
+
+    console.log("stripe payment processor info: ", subscriptions);
```
Another

braydonf commented on lib/models/payment-processor-adapters/stripe.js:
```diff
@@ -0,0 +1,201 @@
+
+    if (subscriptions.length > 1) {
+      // TODO: add actual error handling...
+      console.log('subscription length > 1');
```
Another

braydonf commented on lib/models/payment-processor-adapters/stripe.js:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    if (subscriptions.length < 1) {
+      console.warn('customer has no subscriptions!');
```
Another

braydonf commented on lib/models/payment-processor-adapters/stripe.js:
```diff
@@ -0,0 +1,201 @@
+
+    if (subscriptions.data[0].plan.id !== STRIPE_PLAN_ID) {
+      // TODO: add actual error handling...
+      console.log('stripe plan id wrong');
```
Another

braydonf commented on lib/models/user.js:
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
+    _id: email
+  }, function(err, result) {
     if (err) {
-      return callback(new errors.InternalError(err.message));
```
I think these were already fixed in @barbaraliau pull request, but these should remain as the specific error types.

braydonf commented on lib/models/user.js:
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
-    return callback(
-      new errors.BadRequestError('Password must be hex encoded SHA-256 hash')
-    );
+    return callback(new Error('Password must be hex encoded SHA-256 hash'));
```
Should remain as specific error type

braydonf commented on lib/models/user.js:
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
 
     if (result) {
-      return callback(new errors.BadRequestError('Email is already registered'));
+      return callback(new Error('Email is already registered'));
```
Should remain as specific error type.

braydonf commented on lib/models/user.js:
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
-          return callback(
-            new errors.BadRequestError('Email is already registered')
-          );
+          return callback(new Error('Email is already registered'));
```
Should remain as specific error type

braydonf commented on lib/models/user.js:
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
         }
 
-        return callback(new errors.InternalError(err.message));
+        return callback(err);
```
Should remain as specific error type.

braydonf commented on lib/models/user.js:
```diff
@@ -6,6 +6,129 @@ const activator = require('hat').rack(256);
+PaymentProcessor.methods.update = function(paymentProcessor) {
+  return this.adapter.validate(paymentProcessor)
+    .then((isValid) => {
+      // console.log('this is: ', this);
```
nit: stray commented log

barbaraliau commented:
Starting new clean pull request.
